sync/src/types.rs

```diff
@@ -507,6 +514,69 @@ impl HeaderView {
     pub fn into_inner(self) -> Header {
         self.inner
     }
+
+    pub fn build_skip<F>(&mut self, mut get_header_view: F)
+    where
+        F: FnMut(&H256) -> Option<HeaderView>,
+    {
+        self.skip_hash = get_header_view(self.parent_hash())
+            .and_then(|parent| parent.get_ancestor(get_skip_height(self.number()), get_header_view))
+            .map(|header| header.hash().clone());
+    }
+
+    // NOTE: get_header_view may change source state, for cache or for tests
+    pub fn get_ancestor<F>(self, number: BlockNumber, mut get_header_view: F) -> Option<Header>
+    where
+        F: FnMut(&H256) -> Option<HeaderView>,
+    {
+        let mut current = self;
+        if number > current.number() {
+            return None;
+        }
+        let mut number_walk = current.number();
+        while number_walk > number {
+            let number_skip = get_skip_height(number_walk);
+            let number_skip_prev = get_skip_height(number_walk - 1);
+            match current.skip_hash {
+                Some(ref hash)
+                    if number_skip == number
+                        || (number_skip > number
+                            && !(number_skip_prev + 2 < number_skip
```

Comment:
parent is better if `number_skip_prev < number_skip`, why need `+2` here?

Reply:
`number_skip_prev` and `number_skip` are in different parity(even+odd or odd+even), when `number_skip_prev < number_skip` their distance can be 1, 3, 5...

I guess by current `get_skip_height`, neighbors' parents also neighbors is impossible, so their distance `>=3`

Reply:
`number_skip_prev` and `number_skip` are in different parity (odd+even or even+odd), so their distance can be 1,3,5...

I guess by current `get_skip_height`, neighbor's parents also neighbor is impossible, so their distance `>=3`

Reply:
Need a prove here?